gpu/base/executor.cpp

```diff
@@ -46,34 +46,44 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 namespace gko {
 
 
-void CpuExecutor::raw_copy_to(const GpuExecutor *, size_type num_bytes,
+#define SET_DEVICE_AND_CALL(__device_id__, __function__) \
+    int prev_device;                                     \
+    ASSERT_NO_CUDA_ERRORS(cudaGetDevice(&prev_device));  \
+    ASSERT_NO_CUDA_ERRORS(cudaSetDevice(__device_id__)); \
+    auto errorcode = __function__;                       \
+    ASSERT_NO_CUDA_ERRORS(cudaSetDevice(prev_device))
+
+
+void CpuExecutor::raw_copy_to(const GpuExecutor *dest, size_type num_bytes,
                               const void *src_ptr, void *dest_ptr) const
 {
-    ASSERT_NO_CUDA_ERRORS(
+    SET_DEVICE_AND_CALL(
+        dest->get_device_id(),
         cudaMemcpy(dest_ptr, src_ptr, num_bytes, cudaMemcpyHostToDevice));
+    ASSERT_NO_CUDA_ERRORS(errorcode);
```

Comment:
You are correct, that looks a lot better and works properly. I added the class in the .cpp file directly because there is no other use for it, and in general there should be no other use further on.